swarm/storage/pusher/pusher.go

```diff
@@ -0,0 +1,216 @@
+package pusher
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/chunk"
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+// PushSyncIndex interface
+type PushSyncIndex interface {
+	// subscribe to chunk to be push synced - iterates from earliest to newest
+	SubscribePush(context.Context) (<-chan storage.Chunk, func())
+	// called to set a chunk as synced - and allow it to be garbage collected
+	// TODO this should take ... last argument to delete many in one batch
+	Set(context.Context, chunk.ModeSet, storage.Address) error
+}
+
+// pusher implements push syncing
+type pusher struct {
+	store    PushSyncIndex          // localstore DB
+	quit     chan struct{}          // channel to signal quitting on all loops
+	pushed   map[string]*pushedItem // cache of items push-synced
+	receipts chan chunk.Address     // channel to receive receipts
+	ps       PubSub                 // PubSub interface to send chunks and receive receipts
+}
+
+var (
+	retryInterval = 1 * time.Second // seconds to wait before retry sync
+)
+
+// pushedItem captures the info needed for the pusher about a chunk during the
+// push-sync--receipt roundtrip
+type pushedItem struct {
+	// tag    string    // TODO: tag of the chunk
```

Comment:
Why are all the tag things commented?

Reply:
because it was not done yet. Included now